### ISSUE TITLE
"fragment hints" or "web seeding"

### ISSUE THREAD

**Issue by JamesTheAwesomeDude** [ipfs/ipfs-companion#1049]
**Is your feature request related to a problem? Please describe.**

The "problem" this solves is that of wanting IPFS to serve as a back up for a file against its web host going down, while _simultaneously_ using the web host (or, perhaps, multiple web hosts) as a back up in the case of no-one seeding the file in IPFS.

**Describe the solution you'd like**

Two possible solutions; either/or would be great:

1. The IPFS Companion add-on should attempt to fetch the following file from IPFS, _at least_ in the case of the link itself going HTTP 404/403/410/451:
   https://s3.amazonaws.com/files.cloudprivacy.net/ssl-mitm.pdf#:~:x-ipfs-path=/ipfs/QmVSSCvbYX8XHVcf2kqrpGGmH5PdCbgAP11CCJXrrJQ2yJ
2. The IPFS Companion add-on should attempt to treat Amazon AWS and the other named servers as ["web seed"](https://askubuntu.com/a/313112/95681)-like peers when attempting to fetch the following URL:
   https://ipfs.io/ipfs/QmVSSCvbYX8XHVcf2kqrpGGmH5PdCbgAP11CCJXrrJQ2yJ?filename=ssl-mitm.pdf#:~:x-ipfs-webseed=https%3A//s3.amazonaws.com/files.cloudprivacy.net/ssl-mitm.pdf&x-ipfs-webseed=https%3A//crysp.uwaterloo.ca/courses/cs458/W11-lectures/local/files.cloudprivacy.net/ssl-mitm.pdf&x-ipfs-webseed=https%3A//www.cse.unsw.edu.au/~meyden/3441/ssl-mitm.pdf

**Describe alternatives you've considered**

* Requesting the CDN in question add in `x-ipfs-path` headers to static files
* Hosting it somewhere that includes the appropriate `/ipfs` slug in the path
* Actually seeding the file myself with an IPFS node

**Comment by JamesTheAwesomeDude** [comment]
Obviously, the webseed-hints wouldn't necessarily _need_ to be kept client-side; they could just as well be comprised by a real URL parameter, which might provide some additional flexibility for future improvements.

**Comment by JamesTheAwesomeDude** [comment]
here's another suggestion in the same vein:

```html
this <a href="https://surface.syr.edu/cgi/viewcontent.cgi?article=1846&context=honors_capstone&filename=/tmp/Goppa%20Codes%20and%20Their%20Use%20in%20the%20McEliece%20Cryptosystems.pdf"
   integrity="sha256-nLGeuXDdJZgaa6wpAOSy/81NZ4ngkNjub0ii9tey25c=
              ipfs-mAVUSIJyxnrlw3SWYGmusKQDksv/NTWeJ4JDY7m9IovbXstuX">fantastic explainer by Ashley Valentijn of Syracuse University</a>
```

re ipfs://ipfs/bafkreie4wgpls4g5ewmbu25mfeaojmx7zvgwpcpasdmo432iul3

**Comment by lidel** [comment]
Thank you for brainstorming! Some thoughts / questions (if i missed the mark, please elaborate more):

- **(1)** This flow is tricky, because:
  - data could be imported to IPFS with custom parameters (different chunks, different hash function, different DAG type)
  - CID of data imported with default parameters would not match.
  - Including parameter space in URL is not an option, as one could create DAG by hand, outside of parameters in go-ipfs.
  - The only way to have all the data to deterministically verify a HTTP response is if the web seed is a [CAR](http://ipld.io.ipns.localhost:8080/specs/transport/car/carv1/) file. This is still a good idea :+1: but not sure how useful given someone needs to create a CAR via `ipfs dag export`
- **(2)**  Is the idea here to fetch from IPFS rather than original URL if user agent support IPFS?
  - If so, why not link to https://cf-ipfs.com/ipfs/QmVSSCvbYX8XHVcf2kqrpGGmH5PdCbgAP11CCJXrrJQ2yJ?filename=ssl-mitm.pdf, which will work over HTTP and if ipfs-companion is present get loaded from local node?
- **(3)** iiuc not feasible for various reasons
  - IPFS CID is not a hash of a file, but a hash of a root node of a DAG
  - There is no web extension API that would allow us to do this in a performant way (requires script injection on every  page)
  - If you want content integrity verification, you use `/ipfs/{cid}` paths or DNSLink, and ipfs-companion redirects them to local node, which does the verification natively.

**Comment by JamesTheAwesomeDude** [comment]
> [`/ipfs/QmV…2yJ?filename=…&x-ipfs-webseed=…`] is tricky, because…data could be imported to IPFS with custom parameters (different chunks, different hash function, different DAG type); CID of data imported with default parameters would not match (including parameter space in URL is not an option, as one could create DAG by hand, outside of parameters in go-ipfs).

Torrents also have different chunking parameters that can lead to different URNs for the same exact file (or even fs-tree), and yet solved web-seeding / leeching from "dumb" HTTP servers without provisioning for said parameters in the URI scheme.

While (as with BitTorrent) active/"real"/software peers will be required to bootstrap the _structure_ of the file (i.e. fetching non-leaf data), once the software has constructed a mapping of the chunkspace onto an fs-tree, it's then clear to pull at least the **leaf nodes** from HTTP servers. (Admittedly, a ~0.5MiB PDF wasn't a great prototypical example of this, since most people will be able to snap the whole file up in an instant in the same order of time it takes to fetch the metadata.)

> \[https://cf-ipfs.com/ipfs/QmVSSCvbYX8XHVcf2kqrpGGmH5PdCbgAP11CCJXrrJQ2yJ?filename=ssl-mitm.pdf] will work over HTTP and if ipfs-companion is present get loaded from local node

Per the OP, that (currently) breaks whenever no IPFS client happens to be seeding (pinning) the file at the moment, _even if_ the file's canonical location is still OK. A client supporting `x-ipfs-webseed` parameters would at least have the _possibility_ not to be left in the dark should the seeding peer drop out mid-transfer.

> \[`<a href="…" integrity="ipfs-mAV…tuX">` is] not feasible [because] …There is no web extension API that would allow us to do this in a performant way (requires script injection on every page)

I proposed that _in light of_ [#1050]: given that linkification is seemingly deemed acceptable, the particular rubicon of touching pages isn't a blocker—checking links for a specific attribute should be _far_ more performant than scrubbing all text nodes for excerpts that look like IPFS URIs anyway.
